rundeckapp/grails-app/services/rundeck/services/asyncimport/AsyncImportService.groovy

```diff
@@ -0,0 +1,927 @@
+                        return it
+                    })
+
+                    if (logFound.isPresent()) {
+                        //move it
+                        if( Files.exists(logFound.get()) ){
+                            Files.move(logFound.get(), distributedExecutionsPath.resolve(logFound.get().fileName), StandardCopyOption.REPLACE_EXISTING)
+                        }
+                    }
+                    if (stateFound.isPresent()) {
+                        //move it
+                        if( Files.exists(stateFound.get()) ){
+                            Files.move(stateFound.get(), distributedExecutionsPath.resolve(stateFound.get().fileName), StandardCopyOption.REPLACE_EXISTING)
+                        }
+                    }
+                    // move the execution
+                    if( Files.exists(execution) ){
+                        Files.move(execution, distributedExecutionsPath.resolve(execution.fileName), StandardCopyOption.REPLACE_EXISTING)
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace()
+                throw e
+            }
+
+            deleteNonEmptyDir(tempFile.toString())
+
+            asyncImportStatusFileUpdater(new AsyncImportStatusDTO(projectName, milestoneNumber).with {
+                it.lastUpdate = "Executions distributed, M2 done; proceeding to call M3 event."
+                return it
+            })
+
+            projectService.beginAsyncImportMilestone(
+                    projectName,
+                    authContext,
+                    project,
+                    AsyncImportMilestone.M3_IMPORTING.milestoneNumber
+            )
+
+        }else{
+            asyncImportStatusFileUpdater(new AsyncImportStatusDTO(projectName, milestoneNumber).with {
+                it.lastUpdate = "No executions to iterate, asynchronous import process ended."
+                it.milestoneNumber = AsyncImportMilestone.ASYNC_IMPORT_COMPLETED.milestoneNumber
+                it.milestone = AsyncImportMilestone.ASYNC_IMPORT_COMPLETED.name
+                return it
+            })
+        }
+    }
+
+    /**
+     * Async import process Milestone 3: Files Import
+     * This method is asynchronous and only will be triggered by the completion of milestone 2 through an event emission,
+     * it:
+     * a) Iterates the executions bundles in distributed_executions path of the working directory and for each one:
+     *  1) Copy each bundle to the model project in the working directory
+     *  2) Zip the model project and parse it as an input stream to feed "importToProject"
+     *  3) Calls "importToProject"
+     *  4) Remove the zip and start over until the last bundle is imported.
+     *
+     * @param projectName - String
+     * @param authContext - AuthContext
+     * @param project - IRundeckProject
+     */
+    @Subscriber(AsyncImportEvents.ASYNC_IMPORT_EVENT_MILESTONE_3)
```

Comment:
check if this method can be separated into multiple methods for better reading and easier testing